src/NLog/Targets/AsyncTaskTarget.cs

```diff
@@ -0,0 +1,358 @@
+                _requestQueue.Enqueue(new AsyncLogEventInfo(null, asyncContinuation));
+            }
+        }
+
+        /// <summary>
+        /// Closes Target by updating CancellationToken 
+        /// </summary>
+        protected override void CloseTarget()
+        {
+            _taskTimeoutTimer.Change(Timeout.Infinite, Timeout.Infinite);
+            _cancelTokenSource.Cancel();
+            _requestQueue.Clear();
+            _previousTask = null;
+            base.CloseTarget();
+        }
+
+        /// <summary>
+        /// Releases any managed resources
+        /// </summary>
+        /// <param name="disposing"></param>
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+            if (disposing)
+            {
+                _cancelTokenSource.Dispose();
+                _taskTimeoutTimer.Dispose();
+            }
+        }
+
+        private void TaskStartNext(Task previousTask)
+        {
+            AsyncLogEventInfo logEvent;
+            do
+            {
+                lock (this.SyncRoot)
+                {
+                    if (!IsInitialized)
+                        break;
+
+                    if (previousTask != null && !ReferenceEquals(previousTask, _previousTask))
+                        break;
+
+                    if (_requestQueue.Count == 0)
+                    {
+                        _previousTask = null;
+                        break;
+                    }
+
+                    logEvent = _requestQueue.Dequeue();
+                }
+            } while (!TaskCreation(logEvent));
+        }
+
+        private bool TaskCreation(AsyncLogEventInfo logEvent)
+        {
+            try
+            {
+                if (_cancelTokenSource.IsCancellationRequested)
+                {
+                    logEvent.Continuation(null);
+                    return false;
+                }
+
+                if (logEvent.LogEvent == null)
+                {
+                    InternalLogger.Debug("{0} Flush Completed", this.Name);
+                    logEvent.Continuation(null);
+                    return false;
+                }
+
+                var newTask = WriteAsyncTask(logEvent.LogEvent, _cancelTokenSource.Token);
+                if (newTask == null)
+                {
+                    InternalLogger.Debug("{0} WriteAsyncTask returned null", this.Name);
+                }
+                else
+                {
+                    lock (this.SyncRoot)
+                    {
+                        _previousTask = newTask;
+
+                        if (TaskTimeoutSeconds > 0)
+                            _taskTimeoutTimer.Change(TaskTimeoutSeconds * 1000, Timeout.Infinite);
+
+                        // NOTE - Not using _cancelTokenSource for ContinueWith, or else they will also be cancelled on timeout
+#if (SILVERLIGHT && !WINDOWS_PHONE) || NET4_0
```

Comment:
NET4_5 (and newer) has the new interface with object-parameter, while SILVERLIGHT5 and NET4_0 needs to perform lambda-context-capture.